bot/exts/backend/sync/_syncers.py

```diff
@@ -136,111 +130,67 @@ async def _get_diff(self, guild: Guild) -> _Diff:
-                new_api_user = db_user._replace(
-                    in_guild=False,
-                    **fields_to_none
-                )
-                users_to_update.add(new_api_user)
-
-        new_user_ids = set(guild_users.keys()) - set(db_users.keys())
-        for user_id in new_user_ids:
-            # The user is known on the guild but not on the API. This means
-            # that the user has joined since the last sync. Create it.
-            new_user = guild_users[user_id]
-            users_to_create.add(new_user)
+            def maybe_update(db_field: str, guild_value: t.Union[str, int]) -> None:
+                if db_user[db_field] != guild_value:
+                    updated_fields[db_field] = guild_value
 
-        return _Diff(users_to_create, users_to_update, None)
+            if guild_user := guild.get_member(db_user["id"]):
+                seen_guild_users.add(guild_user.id)
+
+                maybe_update("name", guild_user.name)
+                maybe_update("discriminator", int(guild_user.discriminator))
+                maybe_update("in_guild", True)
 
-    async def _get_users(self, endpoint: str = "bot/users", query_params: list = None) -> t.List[dict]:
-        """GET all users recursively."""
-        users = []
-        response: dict = await self.bot.api_client.get(endpoint, params=query_params)
-        users.extend(response["results"])
+                guild_roles = [role.id for role in guild_user.roles]
+                if set(db_user["roles"]) != set(guild_roles):
+                    updated_fields["roles"] = guild_roles
 
-        # The `response` is paginated, hence check if next page exists.
-        if (next_page_url := response["next"]) is not None:
-            next_endpoint, query_params = self.get_endpoint(next_page_url)
-            users.extend(await self._get_users(next_endpoint, query_params))
-        return users
+            elif db_user["in_guild"]:
+                updated_fields["in_guild"] = False
 
-    @staticmethod
-    def get_endpoint(url: str) -> t.Tuple[str, t.List[tuple]]:
-        """Extract the API endpoint and query params from a URL."""
-        url = urlparse(url)
+            if updated_fields and updated_fields not in users_to_update:
+                updated_fields["id"] = db_user["id"]
+                users_to_update.append(updated_fields)
 
-        # Do not include starting `/` for endpoint.
-        endpoint = url.path[1:]
+        for member in guild.members:
+            if member.id not in seen_guild_users:
+                new_user = {
+                    "id": member.id,
+                    "name": member.name,
+                    "discriminator": int(member.discriminator),
+                    "roles": [role.id for role in member.roles],
+                    "in_guild": True
+                }
+                if new_user not in users_to_create:
```

Comment:
You don't need to check this. I'm pretty sure we can safely assume `guild.members` will not contain duplicates.